src/libstore/local-store.cc

```diff
@@ -221,7 +221,22 @@ LocalStore::LocalStore(ref<const Config> config)
        schema upgrade is in progress. */
     if (!config->readOnly) {
         Path globalLockPath = dbDir + "/big-lock";
-        globalLock = openLockFile(globalLockPath.c_str(), true);
+        try {
+            globalLock = openLockFile(globalLockPath.c_str(), true);
+        } catch (...) {
+            /* It might be more robust to check permissions earlier in this method,
+               but for now take the least invasive approach of providing a better error
+               when the database lock can't be acquired. */
```

Comment:
I wasn't sure whether an explicit check earlier in the method would be less brittle (if e.g. something earlier in the method needs to be root), but that seemed like it had more potential to break something than only changing behavior when an error's already been thrown

Reply:
We can actually check SysError here and see if we get an EPERM error.